dcpy/connectors/edm/packaging.py

```diff
@@ -0,0 +1,150 @@
+import os
+import shutil
+from pathlib import Path
+from typing import Callable
+from dataclasses import dataclass
+import typer
+
+from dcpy.utils import s3
+from dcpy.utils.logging import logger
+from dcpy.connectors.edm import publishing
+
+BUCKET = "edm-distributions"
+BUCKET_ACL = s3.string_as_acl("public-read")
+
+DOWNLOAD_ROOT_PATH = Path(".library")
+OUTPUT_ROOT_PATH = Path(".output")
+
+
+@dataclass
+class PackageKey(publishing.ProductKey):
+    product: str
+    version: str
+
+    def __str__(self):
+        return f"{self.product} - {self.version}"
+
+    @property
+    def path(self) -> str:
+        return f"{self.product}/{self.version}"
+
+
+def colp(publish_key: publishing.PublishKey, package_key: PackageKey):
+    raise NotImplementedError
+
+
+def templatedb(publish_key: publishing.PublishKey, package_key: PackageKey):
+    output_path = OUTPUT_ROOT_PATH / package_key.path
+    shutil.copytree(
+        DOWNLOAD_ROOT_PATH / publish_key.path,
+        output_path,
+        dirs_exist_ok=True,
+    )
+    shutil.copy(
+        output_path / "templatedb.csv",
+        output_path / "templatedb_packaged.csv",
+    )
+
+
+@dataclass
+class PackageMetadata:
+    product: str
+    packaged_name: str
+    packaging_function: Callable
+
+
+DATASET_PACKAGE_METADATA = {
+    "db-colp": PackageMetadata("db-colp", "dcp_colp", colp),
+    "db-template": PackageMetadata("db-colp", "dcp_template_db", templatedb),
+}
+
+
+def upload(package_key: PackageKey) -> None:
+    local_folder_path = OUTPUT_ROOT_PATH / package_key.path
+    meta = s3.generate_metadata()
+    s3.upload_folder(
+        BUCKET,
+        local_folder_path,
+        Path(package_key.path),
+        acl=BUCKET_ACL,
+        contents_only=True,
+        metadata=meta,
+    )
+
+
+def get_packaged_versions(product: str) -> list[str]:
+    return sorted(s3.get_subfolders(BUCKET, f"{product}/"), reverse=True)
+
+
+def transform_for_packaging(
```

Comment:
reason to not just call the passed function?

Reply:
I went back and forth about that. [testing](https://github.com/NYCPlanning/data-engineering/blob/dm-packaging-colp/dcpy/test/connectors/edm/test_packaging.py#L54-L58) led me to isolate this, otherwise a test would have to call `packaging.package` which has a single arg of `publish_key`

to keep the `test-product` inputs in `test/conftest.py` and not define test things in source code (like `packaging.DATASET_PACKAGE_METADATA`), the benefits of modularity seemed worth the cost of having this simple function

Reply:
for 

```python
packaging.transform_for_packaging(
    publish_key,
    package_key,
    mock_data_constants["TEST_PACKAGE_METADATA"].packaging_function,
)
```

can you not do
```python
mock_data_constants["TEST_PACKAGE_METADATA"].packaging_function(publish_key, package_key)
```